### PR TITLE
Abstract over graphql registry atom

### PR DESCRIPTION
**Pull request by frenchy64**
Epic: threatgrid/iroh#3986
Extracted from https://github.com/threatgrid/ctia/pull/924

To port to trapperkeeper, almost all graphql top-level operations will be wrapped in an extra function
that accepts services. To prepare, this PR abstracts over the graphql type registry with a thread-safe function to handle all registry operations.

<a name="qa">[§](#qa)</a> QA
============================

No QA is needed.

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: Abstract over graphql registry atom
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

### CODE REVIEW

File: src/ctia/schemas/graphql/helpers.clj [hunk 1]
```diff
@@ -29,9 +29,27 @@
...
+                                           (not existing-delay)
+                                           (assoc name (delay (f))))))]
+        @result-delay)))
+
 ;; Type registry to avoid any duplicates when using new-object
-;; or new-enum. Contains a map with types indexed by name
-(def default-type-registry (atom {}))
+;; or new-enum. Contains a map with derefable types indexed by name.
+;; Use via get-or-update-type-registry.
+(def ^:private default-type-registry (atom {}))
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:52]: This atom will be created in a Trapperkeeper service in the future.

File: src/ctia/schemas/graphql/helpers.clj [hunk 2]
```diff
@@ -372,17 +395,20 @@
...
-                          (.description description)
-                          (.name union-name)
-                          ; FIXME: this method is deprecated
-                          (.typeResolver type-resolver))]
-    (doseq [type types]
-      (if (instance? GraphQLObjectType type)
-        (.possibleType graphql-union ^GraphQLObjectType type)
-        (.possibleType graphql-union ^GraphQLTypeReference type)))
-    (.build graphql-union)))
+    (get-or-update-type-registry
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:398]: Unions cannot be redefined, as I found out the hard way. As it stands, CTIA doesn't currently doesn't redefined any unions, so the registry wasn't needed.

However, the main strategy of porting to Trapperkeeper is for GraphQL constructors like this one (`new-union`) to return a function that accepts services. Hence, all consumers of a particular GraphQL union will attempt to create it, so the previous invariant is immediately broken for all unions with more than one occurrence in the code base.

File: src/ctia/schemas/graphql/helpers.clj [hunk 3]
```diff
@@ -344,15 +366,16 @@
     interfaces
     fields
     registry]
-   (or (get @registry object-name)
-       (let [builder (-> (GraphQLObjectType/newObject)
-                         (.description description)
-                         (.name object-name)
+    (get-or-update-type-registry
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:369]: Please temporarily permit the funny indentation, a forthcoming commit with wrap this entire return value in a function and restore some order.

File: src/ctia/schemas/graphql/helpers.clj [hunk 4]
```diff
@@ -289,11 +308,14 @@
   [^String object-name
    ^String description
    fields]
-  (-> (GraphQLInputObjectType/newInputObject)
-      (.name object-name)
-      (.description description)
-      (add-input-fields fields)
-      .build))
+    (get-or-update-type-registry
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:311]: Please temporarily permit the funny indentation, a forthcoming commit with wrap this entire return value in a function and restore some order.

File: src/ctia/schemas/graphql/helpers.clj [hunk 5]
```diff
@@ -88,10 +106,12 @@
                                    values
                                    default-type-registry))
   ([^String enum-name ^String description values registry]
-   (or (get @registry enum-name)
-       (let [builder (-> (GraphQLEnumType/newEnum)
+    (get-or-update-type-registry
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:109]: Please temporarily permit the funny indentation, a forthcoming commit with wrap this entire return value in a function and restore some order.

File: src/ctia/schemas/graphql/helpers.clj [hunk 6]
```diff
@@ -289,11 +308,14 @@
...
-      (add-input-fields fields)
-      .build))
+    (get-or-update-type-registry
+      default-type-registry
+      object-name
+      #(-> (GraphQLInputObjectType/newInputObject)
+           (.name ^String object-name)
+           (.description ^String description)
+           (add-input-fields fields)
+           .build)))
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:318]: Please temporarily permit the funny indentation, a forthcoming commit will wrap this entire return value in a function and restore some order.

File: src/ctia/schemas/graphql/helpers.clj [hunk 7]
```diff
@@ -344,15 +366,16 @@
...
-   (or (get @registry object-name)
-       (let [builder (-> (GraphQLObjectType/newObject)
-                         (.description description)
-                         (.name object-name)
+    (get-or-update-type-registry
+      registry
+      object-name
+      #(let [builder (-> (GraphQLObjectType/newObject)
+                         (.description ^String description)
+                         (.name ^String object-name)
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:374]: Please temporarily permit the funny indentation, a forthcoming commit will wrap this entire return value in a function and restore some order.

File: src/ctia/schemas/graphql/helpers.clj [hunk 8]
```diff
@@ -372,17 +395,20 @@
...
+             graphql-union (-> (GraphQLUnionType/newUnionType)
+                               (.description description)
+                               (.name union-name)
+                               ; FIXME: this method is deprecated
+                               (.typeResolver type-resolver))]
+         (doseq [type types]
+           (if (instance? GraphQLObjectType type)
+             (.possibleType graphql-union ^GraphQLObjectType type)
+             (.possibleType graphql-union ^GraphQLTypeReference type)))
+         (.build graphql-union))))
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:411]: Please temporarily permit the funny indentation, a forthcoming commit will wrap this entire return value in a function and restore some order.

File: src/ctia/schemas/graphql/helpers.clj [hunk 9]
```diff
@@ -29,9 +29,27 @@
...
+  (or ;; fast-path for readers
+      (some-> (get @type-registry name) deref)
+      ;; generate a new graphql value, or coordinate with another thread doing the same
+      (let [f (bound-fn* f) ;; may be run on another thread
+            {result-delay name} (swap! type-registry
+                                       (fn [{existing-delay name :as oldtr}]
+                                         (cond-> oldtr
+                                           (not existing-delay)
+                                           (assoc name (delay (f))))))]
+        @result-delay)))
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:47]: This implementation assumes full parallelism, the possibility that `f` may use thread bindings, and an invariant that `f` can only be called once.

File: src/ctia/schemas/graphql/helpers.clj [hunk 10]
```diff
@@ -88,10 +111,12 @@
...
-   (or (get @registry enum-name)
-       (let [builder (-> (GraphQLEnumType/newEnum)
+    (get-or-update-type-registry
+      registry
+      enum-name
+      #(let [builder (-> (GraphQLEnumType/newEnum)
                          (.name enum-name)
-                         (.description description))
+                         (.description ^String description))
              names-and-values? (map? values)]
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:120]: Please temporarily permit the funny indentation, a forthcoming commit will wrap this entire return value in a function and restore some order.

File: src/ctia/schemas/graphql/helpers.clj [hunk 11]
```diff
@@ -88,10 +111,12 @@
                                    values
                                    default-type-registry))
   ([^String enum-name ^String description values registry]
-   (or (get @registry enum-name)
-       (let [builder (-> (GraphQLEnumType/newEnum)
+    (get-or-update-type-registry
+      registry
+      enum-name
+      #(let [builder (-> (GraphQLEnumType/newEnum)
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:117]: Please temporarily permit the funny indentation, a forthcoming commit will wrap this entire return value in a function and restore some order.

File: src/ctia/schemas/graphql/helpers.clj [hunk 12]
```diff
@@ -29,9 +29,33 @@
...
 
-;; Type registry to avoid any duplicates when using new-object
-;; or new-enum. Contains a map with types indexed by name
-(def default-type-registry (atom {}))
+(s/defschema TypeRegistry
+  "Type registry to ensure named GraphQL types are created exactly once.
+  Contains a map with derefable types indexed by name.
+  Use via get-or-update-type-registry."
+  (s/atom {s/Str
+           #_(IDeref graphql.*) ;; returns a graphql-related thing on deref
```
Comment (frenchy64) [inline src/ctia/schemas/graphql/helpers.clj:37]: It's possible this should always be a `graphql.schema.GraphQLType`, but I'm verifying that at the moment.

**Review by ereteog — APPROVED** [review]
LGTM!
I would add a test for `get-or-update-type-registry`.

File: test/ctia/schemas/graphql/helpers_test.clj [hunk 13]
```diff
@@ -1,26 +1,97 @@
...
                            "Description 1"
                            []
-                           {})
-           (sut/new-object "Object"
-                           "Description 2"
-                           []
-                           {})))))
+                           {}
+                           registry)]
+      (is (= (new-object-fn)
```
Comment (vemv) [inline test/ctia/schemas/graphql/helpers_test.clj:22]: Given the `testing "The same object is not created twice"`, you might want to favor `identical?` over `=`

**Review by vemv — APPROVED** [review]
Nice PR!

File: test/ctia/schemas/graphql/helpers_test.clj (hunk 13 above)
Comment (frenchy64) [reply test/ctia/schemas/graphql/helpers_test.clj:22]: Agreed.

**Comment by frenchy64** [comment]
Thanks for the reviews!

**Pull request merged**